### ISSUE TITLE
How to add a custom dropdown aka picker to the quill toolbar?

### ISSUE THREAD

**Issue by oliveratgithub** [quilljs/quill#1817]
Okay so I spent quite some time researching this question and trying various solutions - without success.

**How can I add my custom dropdown menu (aka "picker") to the quill toolbar?**

Here are some - all non-working - approaches I tried so far:

### 1. Using HTML
```html
<div id="toolbar">
	<select class="ql-myDropdown">
		<option value="1">Option 1</option>
		<option value="2">Option 2</option>
		<option value="3">Option 3</option>
		<option value="4">Option 4</option>
		<option value="5">Option 5</option>
	</select>
</div>
<div id="editor">
</div>
```

### 2. Using JavScript in a new Quill instance
```javascript
var quill = new Quill('#editor', {
    modules: {
        toolbar:
        [
          [{ 'header': [1, 2, 3, 4, 5, 6, false] }],
          [{ 'size': ['small', false, 'large', 'huge'] }],
          ['bold', 'italic', 'underline', 'strike',],
          [{ 'myDropdown': ['Option 1', 'Option 2', 'Option 3', 'Option 4', 'Option 5'] }],
    	  ['clean']
        ]
    },
    theme: 'snow'
});
```

### 3. Using a quill import
```javascript
var myDropdown = quill.import('ui/picker');
myDropdown = ['Option 1', 'Option 2', 'Option 3', 'Option 4', 'Option 5'];
quill.register(myDropdown, true);
```

While 2. is the most promising solution - as it does add a picker at the right place and inside Quill; but the heck - it's just an empty dropdown list?

![quill toolbar custom dropdown struggles](https://user-images.githubusercontent.com/280207/32702620-1c88df80-c7ea-11e7-9a3e-16623be6fb59.png)

So honestly I think it would be great, if adding custom dropdowns to Quill would also be explicitely covered in the quill docs. Same as for regular click-buttons.

Thanks for any help!
Oliver

P.s. a related question has been closed without any insights #1543

**Comment by oliveratgithub** [comment]
So a good soul @quannt [over here](https://quannt.github.io/programming/javascript/2017/05/11/adding-custom-toolbar-react-quill.html#comment-3613846315) provided me a jsfiddle demonstrating, what one needs in order to add a custom dropdown picker in Quill: https://jsfiddle.net/q7jferw3/

> Basically quilljs can create the dropdown list for you but you need to manually supply the HTML content (which is not the clearest thing in the world)

Thanks again!!

**Comment by RockStanley** [comment]
Quill uses CSS pseudo class to specify content for each toolbar dropdown menu, so do the menu label.

For example:

``` css
/* static label */
.ql-picker.ql-customoption .ql-picker-label:before {
    content: "Select Language";     /* never changes its content as this menu's label changes */
}
/* dynamic label */
.ql-picker.ql-customoption .ql-picker-label:before[data-value="en"]:before {
    content: "English";
}
/* options */
.ql-picker.ql-customoption .ql-picker-item[data-value="en"]:before {
    content: "English";
}
.ql-picker.ql-customoption .ql-picker-item[data-value="zh"]:before {
    content: "中文";
}
/* you need to specify rules as many as your options */
```

then for the easy part, define the handler:

``` js
new Quill(selector, {
  modules: {
    toolbar: {
      /* ... */
      handlers: {
        'customoption': function(lang: string) {
          /* do something with lang */
        }
      }
    }
  }
}
```

addition: maybe you need to manually specify a padding for the picker label, otherwise it will overlap the pseudo class content.

``` css
.ql-picker.ql-customoption .ql-picker-label {
    padding-right: 18px;
}
```

I traced through the source code to derive this solution.

**Comment by oliveratgithub** [comment]
Sweet, thanks a lot for your efforts @RockStanley !

**Comment by fncamm** [comment]
Thank you @RockStanley , good job!
I've found a way to avoid creating a rule for each option, in my case HTML looks like this 
```html
<select class="ql-custom tooltip" [title]="'values'">
        <option value="1">value1</option>
        <option value="2">value2</option>
        <option value="3">value3</option>
        <option value="4">value4</option>
</select>
```
and the CSS rule is 
```css
.ql-custom .ql-picker-item:before {
  content: attr(data-label);
}
```
The final result looks like this
![image](https://user-images.githubusercontent.com/32817720/44843539-31a49c00-ac51-11e8-8da2-a15fbdffe30d.png)

**Comment by shashwat-vik** [comment]
Just in case if someone comes now and wants a shorter jQuery version for the toolbar styling part given by @oliveratgithub 

```
$.each($('.ql-customoption .ql-picker-item'), function(_, node) {
        node.textContent = node.dataset.value;
});
$('.ql-customoption .ql-picker-label').html('Custom Option' + $('.ql-customoption .ql-picker-label').html());
```

Also for the toolbar to look more perfect and aligned from top I would extend on the styling suggested by @RockStanley
```
// Little more styling for perfect alignment
$('.ql-customoption .ql-picker-label').css('padding-right', '18px');
$('.ql-customoption .ql-picker-label').css('padding-top', '3px');
```

**Comment by allenlongbaobao** [comment]
@oliveratgithub i am sorry for the link https://jsfiddle.net/q7jferw3/ is not working now. Would you please renew one, i do need your help, thanks very much.

**Comment by gbxl** [comment]
You guys rule! Helped quite a bit!

**Comment by KiranVH** [comment]
Hi Guys,

I want to implement my own search functionality on the toolbar button click . Actually what I mean is on button click I want to show a dropdown kind of popup box, inside that I want to give search functionality. So anyone of you know how to proceed for this . please let me know.

Thanks in advance for your time and help

**Comment by T-vK** [comment]
Is anyone interested in a way to dynamically add a dropdown to the toolbar (i.e. without having to provide additional config values during quill instantiation)? 
If there is demand I could provide the code for that.

**Comment by T-vK** [comment]
So I created a library which can do the job: [DynamicQuillTools](https://github.com/T-vK/DynamicQuillTools)

You can use it like this:

``` JavaScript
const dropDownItems = {
    'Mike Smith': 'mike.smith@gmail.com',
    'Jonathan Dyke': 'jonathan.dyke@yahoo.com',
    'Max Anderson': 'max.anderson@gmail.com'
}

const myDropDown = new QuillToolbarDropDown({
    label: "Email Addresses",
    rememberSelection: false
})

myDropDown.setItems(dropDownItems)

myDropDown.onSelect = function(label, value, quill) {
    // Do whatever you want with the new dropdown selection here

    // For example, insert the value of the dropdown selection:
    const { index, length } = quill.selection.savedRange
    quill.deleteText(index, length)
    quill.insertText(index, value)
    quill.setSelection(index + value.length)
}

myDropDown.attach(quill)
```

As you can see, you don't have to mess with the toolbar HTML or the toolbar configuration during the quill instantiation. You can simply add the drop down menu during runtime and even change the values whenever you want, for example you could load them via ajax. 
You can also add a button to the toolbar this way. For more examples, a live demo  and other information follow the link at the top of this post. 

I hope this is useful to someone. 

@KiranVH My library can't really cover your use case yet, but the basis is there and it could be extended to support more complex elements such as the one you're proposing.

**Comment by bgiromini** [comment]
@T-vK Just wanted to say a great big THANK YOU for your nice little library. I had been dreading the idea of hacking quilljs to work with a custom dropdown and had given up. I tried one last time to find some better examples and I landed on your work. It was trivial to implement and works great !

**Comment by abdallahboucedraya** [comment]
how i can implemented in vue quill editor

**Comment by icostaykov** [comment]
Just so if somebody happens to try making a custom dropdown with the "before" CSS pseudocode in Angular - I had huge problems trying to do it. It turns out you need to place `::ng-deep` before all the custom rules, e.g.:
```
::ng-deep .ql-picker.ql-custom .ql-picker-label::before {
  content: "Custom Label";
}

::ng-deep .ql-custom.ql-picker-item::before {
  content: attr(data-value);
}

::ng-deep .ql-picker.ql-custom{
  width: 120px;
}
```

**Comment by yevhenshashnin** [comment]
https://medium.com/@genyadstph/react-quill-add-line-height-to-custom-toolbar-91a8717394c8 this article maybe be helpful if you use react

**Comment by MartirosyanMarat** [comment]
Finally I have found a solution, not the best one, but I hope this is good solution, and **it works for 100%** in my case.

If short **you need add this 2 lines** 



```
::ng-deep .ql-snow .ql-picker.ql-custom-line-height .ql-picker-label[data-label]:not([data-label=""])::before {
  content: attr(data-label);
}

::ng-deep .ql-snow .ql-picker.ql-custom-line-height .ql-picker-item[data-label]:not([data-label=""])::before {
  content: attr(data-label);
}
```



**_Here is my full code_** 

**`component.ts`**

```import {Component, ViewChild} from '@angular/core';
import {QuillEditorComponent, QuillModules, QuillViewComponent, QuillViewHTMLComponent} from "ngx-quill";
import {FormsModule} from "@angular/forms";
import {JsonPipe} from "@angular/common";

@Component({
  selector: 'app-ngx-quill-demo',
  standalone: true,
  imports: [
    QuillEditorComponent,
    QuillViewComponent,
    QuillViewHTMLComponent,
    FormsModule,
    JsonPipe
  ],
  templateUrl: './ngx-quill-demo.component.html',
  styleUrl: './ngx-quill-demo.component.css'
})
export class NgxQuillDemoComponent {
  @ViewChild(QuillEditorComponent, {static: false}) quillEditor!: QuillEditorComponent;

  lineHeight = '1'
  content: any = '<p>Hello World</p>'
  modules: QuillModules = {
    toolbar: {container: '#custom-toolbar'},
  };


  setLineHeight(value: string) {
    if (this.quillEditor) {
      const editorElement = this.quillEditor.editorElem?.querySelector('.ql-editor') as HTMLElement;
      if (editorElement) {
        editorElement.style.lineHeight = value;
      }
    }
  }

  lineHeightChanged() {
    this.setLineHeight(this.lineHeight);
  }
}
```

**`component.html`**

```<quill-editor
  style="width: 100%"
  [style.line-height]="lineHeight"
  [(ngModel)]="content"
  [modules]="modules"
  format="object">
</quill-editor>

<div id="custom-toolbar">
    <span class="ql-formats">
      <button type="button" class="ql-bold"> </button>
      <button type="button" class="ql-italic"> </button>
      <button type="button" class="ql-align" value=""></button>
      <button type="button" class="ql-align" value="center"></button>
      <button type="button" class="ql-align" value="right"></button>
      <button type="button" class="ql-align" value="justify"></button>
    </span>

  <select class="ql-size">
    <option value="small">Small</option>
    <option selected value="normal">Normal</option>
    <option value="large">Large</option>
    <option value="huge">Huge</option>
  </select>

  <!-- Custom Line Height Dropdown -->
  <select class="ql-custom-line-height"
          data-label="'My Line Height'"
          [title]="'Line Height'"
          [(ngModel)]="lineHeight"
          (ngModelChange)="lineHeightChanged()">
    <option value="1" selected>Default</option>
    <option value="1.5">1.5</option>
    <option value="2">2.0</option>
    <option value="2.5">2.5</option>
    <option value="3">3.0</option>
  </select>
</div>
```

**`component.css`**
```::ng-deep .ql-snow .ql-picker.ql-custom-line-height {
  /*This is not must*/
  min-width: 150px;
}

::ng-deep .ql-snow .ql-picker.ql-custom-line-height .ql-picker-label[data-label]:not([data-label=""])::before {
  content: attr(data-label);
}

::ng-deep .ql-snow .ql-picker.ql-custom-line-height .ql-picker-item[data-label]:not([data-label=""])::before {
  content: attr(data-label);
}
```

Note
The only case that I didn't like in this code is that I use the` ::ng-deep`

Wish you the best, I hope this will help you, when I was struggleing I can't fount that will work. 🎉🎉🎉

**Comment by haltingangular-calendar** [comment]
@gbxl https://lnkr.fun/d9b6d9c
